File: proton-c/bindings/cpp/src/session.cpp
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
#include "proton/session.hpp"

#include <proton/connection.h>
#include <proton/session.h>
#include "proton/session.hpp"
#include "proton/connection.hpp"
#include "proton/session_options.hpp"

#include "contexts.hpp"
#include "proton_bits.hpp"

#include <string>

namespace proton {

void session::open() {
    pn_session_open(pn_object());
}

void session::open(const session_options &opts) {
    opts.apply(*this);
    pn_session_open(pn_object());
}

void session::close()
{
    pn_session_close(pn_object());
}

container& session::container() const {
    return connection().container();
}

connection session::connection() const {
    return make_wrapper(pn_session_connection(pn_object()));
}

namespace {
std::string next_link_name(const connection& c) {
    io::link_namer* ln = connection_context::get(c).link_gen;

    return ln ? ln->link_name() : uuid::random().str();
}
}

sender session::open_sender(const std::string &addr) {
    return open_sender(addr, sender_options());
}

sender session::open_sender(const std::string &addr, const sender_options &so) {
    pn_link_t *lnk = pn_sender(pn_object(), next_link_name(connection()).c_str());
    pn_terminus_set_address(pn_link_target(lnk), addr.c_str());
    sender snd(make_wrapper<sender>(lnk));
    snd.open(so);
    return snd;
}

receiver session::open_receiver(const std::string &addr) {
    return open_receiver(addr, receiver_options());
}

receiver session::open_receiver(const std::string &addr, const receiver_options &ro)
{
    pn_link_t *lnk = pn_receiver(pn_object(), next_link_name(connection()).c_str());
    pn_terminus_set_address(pn_link_source(lnk), addr.c_str());
    receiver rcv(make_wrapper<receiver>(lnk));
    rcv.open(ro);
    return rcv;
}

error_condition session::error() const {
    return make_wrapper(pn_session_remote_condition(pn_object()));
}

size_t session::incoming_bytes() const {
    return pn_session_incoming_bytes(pn_object());
}

size_t session::outgoing_bytes() const {
    return pn_session_outgoing_bytes(pn_object());
}

sender_range session::senders() const {
    pn_link_t *lnk = pn_link_head(pn_session_connection(pn_object()), 0);
    while (lnk) {
        if (pn_link_is_sender(lnk) && pn_link_session(lnk) == pn_object())
            break;
        lnk = pn_link_next(lnk, 0);
    }
    return sender_range(sender_iterator(make_wrapper<sender>(lnk), pn_object()));
}

receiver_range session::receivers() const {
    pn_link_t *lnk = pn_link_head(pn_session_connection(pn_object()), 0);
    while (lnk) {
        if (pn_link_is_receiver(lnk) && pn_link_session(lnk) == pn_object())
            break;
        lnk = pn_link_next(lnk, 0);
    }
    return receiver_range(receiver_iterator(make_wrapper<receiver>(lnk), pn_object()));
}

session_iterator session_iterator::operator++() {
    obj_ = pn_session_next(unwrap(obj_), 0);
    return *this;
}

} // namespace proton

File: proton-c/bindings/go/src/qpid.apache.org/electron/session.go
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

package electron

import (
	"qpid.apache.org/proton"
)

// Session is an AMQP session, it contains Senders and Receivers.
type Session interface {
	Endpoint

	// Sender opens a new sender. v can be a string, which is used as the Target
	// address, or a SenderSettings struct containing more details settings.
	Sender(...LinkSetting) (Sender, error)

	// Receiver opens a new Receiver. v can be a string, which is used as the
	// Source address, or a ReceiverSettings struct containing more details
	// settings.
	Receiver(...LinkSetting) (Receiver, error)
}

type session struct {
	endpoint
	eSession   proton.Session
	connection *connection
	capacity   uint
}

// SessionSetting can be passed when creating a sender or receiver.
// See functions that return SessionSetting for details
type SessionSetting func(*session)

// IncomingCapacity sets the size (in bytes) of the sessions incoming data buffer..
func IncomingCapacity(cap uint) SessionSetting { return func(s *session) { s.capacity = cap } }

// in proton goroutine
func newSession(c *connection, es proton.Session, setting ...SessionSetting) *session {
	s := &session{
		connection: c,
		eSession:   es,
		endpoint:   endpoint{str: es.String()},
	}
	for _, set := range setting {
		set(s)
	}
	c.handler.sessions[s.eSession] = s
	s.eSession.SetIncomingCapacity(s.capacity)
	s.eSession.Open()
	return s
}

func (s *session) Connection() Connection     { return s.connection }
func (s *session) eEndpoint() proton.Endpoint { return s.eSession }
func (s *session) engine() *proton.Engine     { return s.connection.engine }
func (s *session) Close(err error) {
	s.engine().Inject(func() { localClose(s.eSession, err) })
}

func (s *session) SetCapacity(bytes uint) { s.capacity = bytes }

func (s *session) Sender(setting ...LinkSetting) (snd Sender, err error) {
	err = s.engine().InjectWait(func() error {
		l, err := localLink(s, true, setting...)
		if err == nil {
			snd = newSender(l)
		}
		return err
	})
	return
}

func (s *session) Receiver(setting ...LinkSetting) (rcv Receiver, err error) {
	err = s.engine().InjectWait(func() error {
		l, err := localLink(s, false, setting...)
		if err == nil {
			rcv = newReceiver(l)
		}
		return err
	})
	return
}

// Called from handler on closed.
func (s *session) closed(err error) {
	s.err.Set(err)
	s.err.Set(Closed)
}

// IncomingSession is passed to the accept() function given to Connection.Listen()
// when there is an incoming session request.
type IncomingSession struct {
	incoming
	h        *handler
	pSession proton.Session
	capacity uint
}

// AcceptCapacity sets the session buffer capacity of an incoming session in bytes.
func (i *IncomingSession) AcceptSession(bytes uint) Session {
	i.capacity = bytes
	return i.Accept().(Session)
}

func (i *IncomingSession) Accept() Endpoint {
	i.accepted = true
	return newSession(i.h.connection, i.pSession, IncomingCapacity(i.capacity))
}
